pkg/logs/sender/destination.go

```diff
@@ -32,26 +33,45 @@ func (e *FramingError) Error() string {
 
 // Destination is responsible for shipping logs to a remote server over TCP.
 type Destination struct {
-	prefixer    Prefixer
-	delimiter   Delimiter
-	connManager *ConnectionManager
-	conn        net.Conn
+	prefixer  Prefixer
+	delimiter Delimiter
+	/* TODO: Merge ConnectionManager into Destination (think about GRPC integration before) */
+	connManager         *ConnectionManager
+	destinationsContext *DestinationsContext
+	conn                net.Conn
 }
 
 // NewDestination returns a new destination.
-func NewDestination(endpoint config.Endpoint) *Destination {
+func NewDestination(endpoint config.Endpoint, destinationsContext *DestinationsContext) *Destination {
 	return &Destination{
-		prefixer:    NewAPIKeyPrefixer(endpoint.APIKey, endpoint.Logset),
-		delimiter:   NewDelimiter(endpoint.UseProto),
-		connManager: NewConnectionManager(endpoint),
+		prefixer:            NewAPIKeyPrefixer(endpoint.APIKey, endpoint.Logset),
+		delimiter:           NewDelimiter(endpoint.UseProto),
+		connManager:         NewConnectionManager(endpoint),
+		destinationsContext: destinationsContext,
 	}
 }
 
 // Send transforms a message into a frame and sends it to a remote server,
 // returns an error if the operation failed.
 func (d *Destination) Send(payload message.Message) error {
 	if d.conn == nil {
-		d.conn = d.connManager.NewConnection()
+		var ctx context.Context
+		var err error
+
+		// If we have a context, use it, this will allow early cancellation.
+		if d.destinationsContext != nil {
```

Comment:
as discussed IRL, it seemed that you had a better idea to tackle this.

Reply:
This was for `d.destinationsContext.Context() = nil` (after a `Stop()`), which I did.
This check is still necessary for make the `d.destinationsContext` optional (unit tests)